### PR TITLE
Add compiler pragmas to disable -Wfloat warnings locally

### PR DESCRIPTION
**Pull request by phillipjohnston**
-Wfloat warnings are valuable, but in this library float equality is expected and fine within the context. Compiler preprocessor steps have been added to disable this warning for GCC and Clang.

These pragmas can be disabled completely by defining PRINTF_DISABLE_WFORMAT_WARNING_PRAGMAS

### CODE REVIEW

File: src/printf/printf.c [hunk 1]
```diff
@@ -78,6 +78,17 @@ extern "C" {
 # define vprintf_   vprintf
 #endif
 
+// If you have -Wfloat-equal enabled, this source code will emit a number
+// of warnings for float comparisons for equality. However, within the
+// context of their use in this library, these are ok to use. So we
+// disable the warning for this file.
+// Define PRINTF_DISABLE_WARNING_PRAGMAS to disable this block
+#ifndef PRINTF_DISABLE_WARNING_PRAGMAS
```
Comment (eyalroz) [inline src/printf/printf.c:86]: Why do I want to make this conditional? Isn't conditioning on `__GNUC__` or `__clang__` enough?

**Review by eyalroz — CHANGES_REQUESTED** [review]

File: src/printf/printf.c (hunk 1 above)
Comment (phillipjohnston) [reply src/printf/printf.c:86]: Just trying to enable flexibility - maybe there's a reason you want to compile with those warnings enabled.

**Comment by eyalroz** [comment]
I still don't know whether this PR makes sense. I mean, we don't build `printf.c` with -Wfloat-equal; and that flag is not turned on by default in any compiler I know of, not even with gcc's `-Wextra` . So why should I insert preprocessor code to protect against that? Should it not be the responsibility of whoever decides to build `printf.c` with this flag? If they're altering the build system anyway, why don't they just make an exception for `printf.c`?

**Comment by phillipjohnston** [comment]
I would be more amenable to that argument if you did not already use such pragmas in the code base.

**Comment by phillipjohnston** [comment]
That being said, I completely understand not accepting the change. The main impact on me is that I will just maintain a separate fork of this project for all time. That is much easier than than completely redesigning an existing build system for a libc project to avoid one compiler flag on one file.

**Comment by eyalroz** [comment]
> I would be more amenable to that argument if you did not already use such pragmas in the code base.

Where? You mean, in the test suite? The test suite is something else. That's not installed. That's not code people need to read through...

> The main impact on me is that I will just maintain a separate fork of this project for all time. That is much easier than than completely redesigning an existing build system for a libc project to avoid one compiler flag on one file. (Our build system does not support per-source-file arguments).

I haven't closed this PR yet...

**Comment by phillipjohnston** [comment]
I don't expect all of my code to not perform floating-point comparisons, I just generally follow the advice of avoiding `==` because of the [problems with comparing floating point numbers](https://bitbashing.io/comparing-floats.html) (which is why a warning exists). This has bitten me more than once on a production system. There are much better alternatives than `==` - for example, using an epsilon (although selecting the right epsilon value is also fraught with difficulty, as are epsilons around zero) and comparing them for equality using integers instead of floats.

**Comment by phillipjohnston** [comment]
Now, as to me closing the PR... you already indicated you were disinclined to accept it, and I did not feel as if my argument would be sufficiently convincing, so I decided to just proceed in a different direction rather than expend further effort convincing you. Your re-opening comment mostly contains other reasons you're disinclined to accept it, so that still doesn't seem like I misinterpreted the odds :).

**Comment by eyalroz** [comment]
Ok... sorry :-(

I hope the `printf_config.h` route could help you avoid a bunch of work duplication.

**Comment by phillipjohnston** [comment]
It’s a good suggestion and I will definitely explore it as an option

**Pull request closed without merging**